colossalai/inference/tensor_parallel/utils.py

```diff
@@ -0,0 +1,46 @@
+import types
+
+import torch
+from vllm.model_executor.input_metadata import InputMetadata
+from vllm.model_executor.models.llama import LlamaAttention, LlamaForCausalLM, LlamaModel
+
+from colossalai.inference.continous_batching.layers.attention import PagedAttentionWithRoPE
+
+
+def init_to_get_rotary(self, base=10000):
+    self.config.head_dim_ = self.config.hidden_size // self.config.num_attention_heads
+    if not hasattr(self.config, "rope_scaling"):
+        rope_scaling_factor = 1.0
+    else:
+        rope_scaling_factor = self.config.rope_scaling.factor if self.config.rope_scaling is not None else 1.0
+    if hasattr(self.config, "max_sequence_length"):
+        max_seq_len = self.config.max_sequence_length
+    elif hasattr(self.config, "max_position_embeddings"):
+        max_seq_len = self.config.max_position_embeddings * rope_scaling_factor
+    else:
+        max_seq_len = 2048 * rope_scaling_factor
+    base = float(base)
+    inv_freq = 1.0 / (base**(torch.arange(0, self.config.head_dim_, 2, device="cpu", dtype=torch.float32) /
+                             self.config.head_dim_))
+    t = torch.arange(max_seq_len + 1024 * 64, device="cpu", dtype=torch.float32) / rope_scaling_factor
+    freqs = torch.outer(t, inv_freq)
+
+    self._cos_cached = torch.cos(freqs).to(torch.float16).cuda()
+    self._sin_cached = torch.sin(freqs).to(torch.float16).cuda()
+    return
+
+
+def replace_model(model, kv_cache_stream):
```

Comment:
Might modify the name of the function (a little bit confusing, since the function focuses on replacing attention layer with paged attention,

Reply:
Done